### PR TITLE
Check user merging privilege

### LINKED ISSUE #176: Oppiabot does not assign author after pull request has been approved
In some pull requests, oppiabot does not assign the appropriate user to merge the pull request even after adding the LGTM label.

See https://github.com/oppia/oppia/pull/11376, https://github.com/oppia/oppia/pull/11316.

__Expected Behaviour__

Oppiabot should assign the PR author if they have merging privileges or assign the project owner from the changelog label to merge the pull request if the PR author doesn't have merging privileges.

### PR DESCRIPTION
**Pull request by riyasingh1004**
## Explanation
Fixes #176  Added  a check for merge permissions of a user. Ater LGTM label is added oppiabot checks for appropriate user who can be assigned to merge the PR.

## Checklist
- [ ] I have successfully deployed my own instance of Oppiabot.
  - You can find instructions for doing this [here](https://github.com/oppia/oppiabot/wiki/Deploying-your-own-instance-of-the-oppiabot).
- [ ] I have manually tested all the changes made in this PR following the [manual tests matrix](https://github.com/oppia/oppiabot/wiki/Manual-Tests-Matrix).

### CODE REVIEW

**Comment by riyasingh1004** [comment]
@jameesjohn Please review. I am not sure how to correctly check for admin and write permissions, that's why the checks are failing.

File: lib/utils.js [hunk 1]
```diff
@@ -17,6 +17,7 @@
  */
 
 const { 'default': Axios } = require('axios');
+const { admin } = require('googleapis/build/src/apis/admin');
```
Comment (jameesjohn) [inline lib/utils.js:20]: Why is this imported?

File: lib/utils.js [hunk 2]
```diff
@@ -240,6 +241,28 @@ const isUserCollaborator = async (context, username) => {
...
+ * @param {string} username
+ */
+
+const hasMergePrivilege = async (context, username) => {
+  try {
+    const collaboratorPrivilege = await context.github.repos.getCollaboratorPermissionLevel({
+      owner: context.repo().owner,
+      repo: context.repo().repo,
+      username: username
+    });
```
Comment (jameesjohn) [inline lib/utils.js:257]: To test this, you can log the result from this call, and inspect the outputs.

**Review by jameesjohn — CHANGES_REQUESTED** [review]
Thanks, @riyasingh1004, I left some comments.

File: lib/utils.js (hunk 1 above)
Comment (riyasingh1004) [reply lib/utils.js:20]: This was a mistake, I will fix it in the next commit.

**Comment by riyasingh1004** [comment]
@jameesjohn PTAL,  I fixed the permission check and also tried to update the tests but I am still confused what all changes are still required to fix the tests.

**Comment by riyasingh1004** [comment]
user-check1

**Comment by riyasingh1004** [comment]
@jameesjohn Can you help me where the checks are failing. I got the user permission level as read, write or admin by using data.permission from the above response. So I guess there is some error due to that in the spec file.

**Review by jameesjohn — CHANGES_REQUESTED** [review]
Thanks, @riyasingh1004, I took another pass. 

The lint checks are failing primarily because some lines have lengths greater than 80 characters.

For the failing tests, you need to mock `github.repos.getCollaboratorPermissionLevel` in the test file. Similar to how other functions are mocked there.

See https://github.com/oppia/oppiabot/blob/959b59c9ccc6cd1494b249f6b1fa11176258bbfc/spec/checkPullRequestReviewSpec.js#L691-L700

File: lib/utils.js [hunk 3]
```diff
@@ -240,6 +240,29 @@ const isUserCollaborator = async (context, username) => {
...
+/**
+ * Check if the user has merge permissions.
+ *
+ * @param {import('probot').Context} context
+ * @param {string} username
+ */
+
+const hasMergePrivilege = async (context, username) => {
+  try {
+    const collaboratorPrivilege = await context.github.repos.getCollaboratorPermissionLevel({
```
[the remark this answers is not in the document]
Comment (riyasingh1004) [reply lib/utils.js]: Done

File: lib/utils.js [hunk 4]
```diff
@@ -240,6 +240,29 @@ const isUserCollaborator = async (context, username) => {
...
+      owner: context.repo().owner,
+      repo: context.repo().repo,
+      username: username
+    });
+    return (collaboratorPrivilege.data.permission === 'admin') || (collaboratorPrivilege.data.permission === 'write');
```
[the remark this answers is not in the document]
Comment (riyasingh1004) [reply lib/utils.js]: Done

**Comment by riyasingh1004** [comment]
@jameesjohn PTAL , Regarding the tests I replaces the should PR author can merge checks . Should I restore those checks and add another check for author merging privilege??
Screenshot from 2021-03-29 23-50-00

**Review by jameesjohn — CHANGES_REQUESTED** [review]
Hi @riyasingh1004!
Apologies for the late review.

The lint tests are failing mostly because some lines are over 80 characters long.
The frontend tests are failing because `github.repos.getCollaboratorPermissionLevel` has not been mocked in places where they are bing used. Like https://github.com/oppia/oppiabot/blob/a9c06dc340450a5d40cd3aa42bebaec421650b7c/spec/periodicChecksSpec.js#L154-L168

**Comment by riyasingh1004** [comment]
@jameesjohn PTAL.

File: lib/utils.js [hunk 5]
```diff
@@ -240,6 +241,30 @@ const isUserCollaborator = async (context, username) => {
...
+ * @param {string} username
+ */
+
+const hasMergePrivilege = async (context, username) => {
+  try {
+    const collaboratorPrivilege = await context.github.repos
+    .getCollaboratorPermissionLevel({
+      owner: context.repo().owner,
+      repo: context.repo().repo,
+      username: username
```
[the remark this answers is not in the document]
Comment (jameesjohn) [reply lib/utils.js:257]: Not done?

File: spec/checkPullRequestReviewSpec.js [hunk 6]
```diff
@@ -804,10 +807,11 @@ describe('Pull Request Review Module', () => {
...
               }),
           };
-          github.orgs = {
-            checkMembership: jasmine.createSpy('checkMembership')
+          github.repos = {
+            getCollaboratorPermissionLevel: jasmine
+            .createSpy('getCollaboratorPermissionLevel')
               .and.resolveTo({
-                status: 204,
+                status: 200,
```
Comment (jameesjohn) [inline spec/checkPullRequestReviewSpec.js:814]: This should resolve to an object with the following structure `{data:{status:<status>, permission<permission>}}`

**Review by jameesjohn — CHANGES_REQUESTED** [review]
You should also modify the periodic checks spec since it used the previous implementation.

File: spec/checkPullRequestReviewSpec.js (hunk 6 above)
Comment (riyasingh1004) [reply spec/checkPullRequestReviewSpec.js]: Done

File: lib/utils.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (riyasingh1004) [reply lib/utils.js]: Done

**Comment by riyasingh1004** [comment]
@jameesjohn PTAL

**Comment by riyasingh1004** [comment]
@jameesjohn PTAL

**Comment by jameesjohn** [comment]
No need to update the above as it was done in https://github.com/oppia/oppiabot/pull/270.
So updating this branch with master would fix the tests.

**Comment by riyasingh1004** [comment]
@jameesjohn  PTAL

**Review by jameesjohn — APPROVED** [review]
LGTM!

**Pull request merged**